osg-import-srpm

```diff
@@ -75,106 +86,326 @@ def make_svn_tree(srpm, url, extract_spec=False, diff_spec=False, provider=None)
+                provider = provname
+                break
         else:
-            provider_name = 'developer'
-    else:
-        provider_name = provider
+            provider = 'developer'
 
-    source_filename = opj(upstream_dir, provider_name+".srpm.source")
+    source_filename = os.path.join(upstream_dir, provider+".srpm.source")
 
-    source_contents = "%s\n# Downloaded from '%s'\n" % (opj(name, version, srpm), url)
+    source_contents = """\
+%(cached_filename)s
+# Downloaded from '%(url)s'
+""" % locals()
 
-    if exists(source_filename):
-        print "Source file exists. Saving as %s.old" % source_filename
-        move(source_filename, source_filename + ".old")
-        unslurp(source_filename, source_contents)
+    if os.path.exists(source_filename):
+        logging.info("Source file exists. Saving as %s.old", source_filename)
+        shutil.move(source_filename, source_filename + ".old")
+        utils.unslurp(source_filename, source_contents)
     else:
-        unslurp(source_filename, source_contents)
-        # TODO Check the path we're trying to add hasn't been added yet
-        checked_call(["svn", "add", source_filename])
-    if len(glob.glob(opj(upstream_dir, "*.source"))) > 1:
-        print "More than one .source file found in upstream dir."
-        print "Examine them to make sure there aren't duplicates."
-
-    osg_dir = opj(name, "osg")
-    if diff_spec:
-        # TODO: This needs touchups.
-        if not isdir(osg_dir):
-            print "No osg dir exists. Not diffing."
+        utils.unslurp(source_filename, source_contents)
+        svn_safe_add(source_filename)
+
+
+def is_untracked_path(path):
+    """Return True if the given path is untracked in SVN.
+    Note: ignored files return False.
+    """
+    output, ret = utils.sbacktick(["svn", "status", path])
+
+    return output.startswith('?')
+
+
+def svn_safe_add(path):
+    """Add path to SVN if it's not already in there. Return True on success."""
+    if is_untracked_path(path):
+        ret = utils.unchecked_call(["svn", "add", path])
+        return ret == 0
+
+
+def get_spec_name_in_srpm(srpm):
+    """Return the name of the spec file present in an SRPM.  Assumes
+    there is exactly one spec file in the SRPM -- if there is more than
+    one spec file, returns the name of the first one ``cpio'' prints.
+    """
+    out, ret = utils.sbacktick("rpm2cpio %s | cpio -t '*.spec' 2> /dev/null" % re.escape(srpm), shell=True)
+    if ret != 0:
+        raise Error("Unable to get list of spec files from %s" % srpm)
+    try:
+        spec_name = filter(None, [x.strip() for x in out.split("\n")])[0]
+    except IndexError:
+        spec_name = None
+
+    if not spec_name:
+        raise Error("No spec file inside %s" % srpm)
+
+    return spec_name
+
+
+def extract_from_rpm(rpm, file_or_pattern=None):
+    """Extract a specific file or glob from an rpm."""
+    command = "rpm2cpio " + re.escape(rpm) + " | cpio -ivd"
```

Comment:
Here as in a few other places, note that `re.escape` does not do the right thing when quoting filenames with embedded newlines.  `pipes.quote` seems to work for python 2.4 & 2.6.  (`shlex.quote` in 2.7+)